src/imperative/batching.cc

```diff
@@ -59,28 +63,132 @@ nnvm::Graph DBatchEngine::BatchGraphs(std::vector<nnvm::Graph>& graphs) {
-  // (TODO szha)
+  // create mapping from old node to new node
+  std::unordered_map<nnvm::Node*, std::vector<nnvm::NodePtr>> old_new_node_map;
+  for (uint32_t istep = 0; istep < forward_steps.size(); istep++) {
+    const std::unordered_map<int64_t, std::vector<nnvm::NodePtr>> step = forward_steps[istep];
+    for (const std::pair<int64_t, std::vector<nnvm::NodePtr>> step_ops : step) {
+      int64_t op_sign = step_ops.first;
+      const std::vector<nnvm::NodePtr> op_ptrs = step_ops.second;
+      size_t num_nodes = op_ptrs.size();
+      nnvm::Node* first_op_node = op_ptrs.front().get();
+      if (num_nodes == 1) { // op that can't be batched
+        std::vector<nnvm::NodeEntry> old_inputs = first_op_node->inputs;
+        std::vector<nnvm::NodeEntry> new_inputs(old_inputs.size());
+        for (const nnvm::NodeEntry input_entry : old_inputs) {
+          new_inputs.emplace_back(map_node_entry(input_entry,
+                                                 old_new_node_map));
+        }
+        nnvm::NodePtr new_node = nnvm::Node::Create();
+        new_node->attrs = first_op_node->attrs;
+        new_node->inputs = new_inputs;
+        CHECK_EQ(first_op_node->control_deps.size(), 0); // TODO map control_deps node ptr
+        new_node->info = first_op_node->info;
+        old_new_node_map[first_op_node].emplace_back(new_node);
+        continue;
+      }
+
+      size_t num_inputs = first_op_node->num_inputs(), num_outputs = first_op_node->num_outputs();
+
+      std::unordered_set<size_t> batchable_data_indices = {0}; // TODO find batchable data tensors from op
 
+      // record lengths of each sample
+      std::vector<size_t> sample_lengths(num_nodes);
+      for (uint32_t j = 0; j < num_nodes; j++) {
+        sample_lengths[j] = 1; // TODO get shape from node->ndarray mapping
+      }
+
+      std::vector<nnvm::NodeEntry> batch_node_inputs(num_inputs);
+      for (uint32_t i = 0; i < num_inputs; i++) {
+        if (batchable_data_indices.find(i) != batchable_data_indices.end()) {
+          // batchable input, create concat node
+          size_t in_batch_axis = 0; // TODO get batch axis from nodes for each input index
+
+          std::vector<nnvm::NodeEntry> concat_node_inputs(num_nodes);
+          std::string concat_node_name = "batch_input_" + std::to_string(i) + "_concat";
+          for (const nnvm::NodePtr op_ptr : op_ptrs) {
+            nnvm::NodeEntry input_entry = op_ptr.get()->inputs[i];
+            concat_node_inputs.emplace_back(map_node_entry(input_entry,
+                                                           old_new_node_map));
+            concat_node_name += "_" + input_entry.node->attrs.name;
+          }
+          std::unordered_map<std::string, std::string> concat_args = {
+            {"num_args", std::to_string(num_nodes)},
+            {"dim", std::to_string(in_batch_axis)}
+          };
+
+          batch_node_inputs.emplace_back(MakeNode("concat", "concat_", concat_node_inputs,
```

Comment:
maybe move CreateConcatNode outside as a function

Reply:
also add cache for batched input

Reply:
cache look-up can be expensive. will not fix.